src/plugins/kibana/server/routes/api/ingest/register_bulk.js

```diff
@@ -0,0 +1,98 @@
+import { Promise } from 'bluebird';
+import { parse, transform } from 'csv';
+import _ from 'lodash';
+import hi from 'highland';
+import { patternToIngest } from '../../../../common/lib/convert_pattern_and_ingest_name';
+import { PassThrough } from 'stream';
+import bulkRequestSchema from '../../../lib/schemas/bulk_request_schema';
+
+export function registerBulk(server) {
+  server.route({
+    path: '/api/kibana/{id}/_bulk',
+    method: 'POST',
+    config: {
+      payload: {
+        output: 'stream',
+        maxBytes: 1024 * 1024 * 1024
+      },
+      validate: {
+        payload: bulkRequestSchema
+      }
+    },
+    handler: function (req, reply) {
+      const boundCallWithRequest = _.partial(server.plugins.elasticsearch.callWithRequest, req);
+      const indexPattern = req.params.id;
+      const usePipeline = req.payload.pipeline;
+      const csv = req.payload.csv;
+      const fileName = csv.hapi.filename;
+      const responseStream = new PassThrough();
+      const parser = parse({
+        columns: true,
+        auto_parse: true,
+        delimiter: _.get(req.payload, 'delimiter', ',')
+      });
+
+      let currentLine = 2; // Starts at 2 since we parse the header separately
+
+      responseStream.write('[');
+      csv.pipe(parser);
+
+      hi(parser)
```

Comment:
First off, I'll say that my intent isn't for this initial version to be a bullet proof API that third parties can rely on, it just needs to work for the CSV upload UI. I made the mistake of trying to build a general purpose API from the get go with the ingest API, and that led to a ton of churn and wasted time. So it might not handle every single possible edge case, but it doesn't need to.

I'm not sure how to respond to the rest of your comment since you just wrote a list of vocab words instead of providing actionable feedback. But I'll try to address each item you listed:

* size of batching - it's handled by highland. I don't generally write unit tests for third party code, do you? Also the size of batching is an implementation detail, not something that should be tested as a part of the public API
* amount of parallelization - it's handled by highland. Again, this is an implementation detail
* back pressure management - it's handled by highland and node. This is the entire reason I used highland instead of writing my own code to handle reading the csv parser stream and sending requests to elasticsearch 
* authorization - handled by `boundCallWithRequest`, which should have its own tests. Also easy to test with functional tests, send a request with bad credentials, check for 403. I'll admit, this currently isn't being done.
* base path handling - Like the one we add with the base path proxy? I don't see how that's relevant given that this API returns no links, but you could write functional tests with a proxy involved to verify it works.
* handling multi-part csv request bodies - All of the functional tests send multipart/form-data request bodies, and they're all passing, so that seems to be working. If the API suddenly stopped supporting multipart/form-data, every test would fail.
* rejecting bad requests - See the functional tests for parse errors, indexing errors, and a test verifying a 400 is returned if the one required parameter is missing.
* handling csv-specific request parameters like delimiter - Yup, got functional tests for that. For both `delimiter` and `pipeline`.
* streaming responses - I have a functional test that verifies the response is sent with chunked encoding. See `should stream a chunked response`. And if you're talking about verifying that the streamed response is valid json, see every test that has an assertion on the response body.
* etc - I'm not a mind reader, but if you have additional concerns I'm happy to address them.

So there you go... I didn't just show you *how* all those things could be verified with functional tests alone, but how I'm *already* doing it.